src/clj/fluree/db/reasoner.cljc

```diff
@@ -0,0 +1,276 @@
+(ns fluree.db.reasoner
+  (:require [clojure.core.async :as async :refer [alts! go]]
+            [clojure.string :as str]
+            [fluree.db.flake :as flake]
+            [fluree.db.query.fql.parse :as parse]
+            [fluree.db.util.core :as util :refer [try* catch*]]
+            [fluree.db.util.log :as log]
+            [fluree.db.util.async :refer [go-try <?]]
+            [fluree.db.reasoner.resolve :as resolve]
+            [fluree.db.json-ld.transact :as transact]
+            [fluree.db.fuel :as fuel]
+            [fluree.db.constants :as const]
+            [fluree.json-ld :as json-ld]
+            [fluree.db.query.fql.parse :as q-parse]
+            [fluree.db.reasoner.owl-datalog :as owl-datalog]
+            [fluree.db.dbproto :refer [db?]]
+            [fluree.db.reasoner.graph :refer [task-queue add-rule-dependencies]]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn reasoned-rule
+  "Returns truthy if the flake has been generated by reasoner"
+  [flake]
+  (-> flake meta :reasoned))
+
+(defn non-reasoned-flakes
+  "Takes a sequence of flakes and removes any flakes which are reasoned.
+
+  This is primarily used to remove reasoned flakes from commits."
+  [flakes]
+  (remove reasoned-rule flakes))
+
+(defn reasoned-flakes
+  "Takes a sequence of flakes and keeps only reasoned flakes"
+  [flakes]
+  (filter reasoned-rule flakes))
+
+(defn schedule
+  "Returns list of rule @id values in the order they should be run.
+
+  If optional result-summary is provided, rules that don't need to be
+  re-run will be filtered out.
+
+  A result summary is list/set of the rule dependency patterns which
+  match newly created Flakes from the last run. When the result-summary
+  is empty, no rules will be left to run, but based on the dependencies
+  it is possible no rules will be left to run even if the result-summary
+  is non-empty"
+  ([rules]
+   (task-queue rules))
+  ([rules result-summary]
+   (task-queue rules result-summary)))
+
+(defn reasoner-insert
```

Comment:
This function and `reasoner-stage` have a lot of logic in common. I think one of them could be defined in terms of the other to avoid duplicating that logic.

Reply:
Addressed in commit beb17dd217df5a5dfd146907c14fcea7a8f63529